js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts

```diff
@@ -0,0 +1,242 @@
+import { getId, getRichLabel, isIdentifiable, isLabeled } from '@prestojs/util';
+
+/**
+ * A single choice
+ */
+export interface Choice<T> {
+    value: T;
+    label: React.ReactNode;
+    [x: string]: any;
+}
+/**
+ * Grouped choices - a label for the grouping and an array of choices for that label
+ */
+export type ChoicesGrouped<T> = [string, Choice<T>[]];
+
+/**
+ * Interface for asynchronous choices.
+ *
+ * A choice is a `value` that identifies the item (eg. an id) and a `label` that
+ * describes the item and is shown to users (eg. the name).
+ *
+ * When `multiple` is true multiple values can be selected.
+ *
+ * To define async choices two things are required:
+ *
+ * 1) A function to resolve existing value(s). This is used when viewing existing
+ *    value(s) and need label(s) to show (eg. when displaying a choice on a detail
+ *    view or rendering the value selected on a select widget).
+ * 2) A function to list & filter the available  choices. This is used when selecting
+ *    a value (eg. the options shown in a select widget).
+ *
+ * Both of these functions may need to store state (eg. pagination for a listing)
+ * or access things from context (eg. read values from a cache). This can be done
+ * via two hooks - `useListProps` and `useRetrieveProps`. This functions should be
+ * called from a component or hook that deals with async choices when calling
+ * `list` and `retrieve` respectively. The return value from the hook is passed to
+ * the corresponding function.
+ *
+ * For a default implementation see [AsyncChoices](doc:AsyncChoices)
+ *
+ * @extract-docs
+ * @menu-group Async Choices
+ */
+export interface AsyncChoicesInterface<ItemType, ValueType> {
+    /**
+     * If true then multiple values can be selected. When this is true
+     * retrieve() will be passed and return an array rather than a single value.
+     */
+    multiple: boolean;
+    /**
+     * Hook that returns dependencies for listing choices. This is useful to store state
+     * for things like pagination. The value returned here is passed to `list`.
+     *
+     * What this function is passed depends on the implementation but when used with
+     * [useAsyncChoices](doc:useAsyncChoices) it will be passed `value` (the currently
+     * selected value), `query` (the query object eg. to filter results with) and `listOptions`
+     * (any additional options passed on the `listOptions` prop to `useAsyncChoices`).
+     * Official presto widgets all use `useAsyncChoices` and so will use these parameters.
+     */
+    useListProps(args: any): any;
+    /**
+     * Function to resolve a list of choices based on the provided params.
```

Comment:
`params` is where an autocomplete keyword search string would fit in? How would that get injected given that `useListProps()`is passed straight in?

Reply:
See https://github.com/prestojs/prestojs/pull/67/files#diff-a1b1883790df4d0587b34a90efa4a851R120

You can return it from `listProps` otherwise it's passed straight through in `useAsyncChoices`